### PR TITLE
Fixed and clean some of some code after test

### PR DESCRIPTION
**Pull request by mossicrue**

### CODE REVIEW

File: katello-attach-subscription [hunk 1]
```diff
@@ -907,13 +896,37 @@ def vdcupdate()
       else
         virtualhostregex = nil
       end
+      # test all the facts
+      if @options[:verbose]
+        puts "VERBOSE: Start testing facts value for #{sys['name']}"
+      end
+      fact_test_passed = compareAllFacts(sub, sys)
```
Comment (evgeni) [inline katello-attach-subscription]: why don't you use `KatelloAttachSubscription::HostMatcher.match_host(sys, sub)` here?

File: katello-attach-subscription [hunk 2]
```diff
@@ -1455,7 +1479,8 @@ end
 # count the total number of subs available, consumed and free
 def getsubdetailfor(sub_name)
   detail_hash = {"consumed" => 0, "free" => 0}
-  search_options = KatelloAttachSubscription::Utils.search_args(name: sub_name)
+  search_params = 'name="'+sub_name+'"'
+  search_options = KatelloAttachSubscription::Utils.search_args(search_params)
```
Comment (evgeni) [inline katello-attach-subscription]: this will make `search_options` `"name="the_subname""` which sounds wrong to me?

File: katello-attach-subscription [hunk 3]
```diff
@@ -433,14 +437,14 @@ def checksubs()
...
+            puts "VERBOSE: adding virt-who data to #{sys['name']}"
           end
-          sys = KatelloAttachSubscription::VirtWhoHelper.merge_system_virtwho(sys, @parsed_hypervisors_hash, @options[:organization_id])
+          sys = KatelloAttachSubscription::VirtWhoHelper.merge_system_virtwho(sys, @parsed_hypervisors_hash, @options[:org])
         end
         hypervisors_collection.push(sys)
-      rescue RestClient::NotFound
-        STDERR.puts "   ERROR: Hypervisor '#{system['name']}' not found. Skipping."
-        next
+      #rescue RestClient::NotFound
```
Comment (evgeni) [inline katello-attach-subscription]: why?

File: katello-attach-subscription [hunk 4]
```diff
@@ -433,14 +437,14 @@ def checksubs()
         sys = @api.resource(:hosts).call(:show, {:id => system['id'], :fields => 'full'})
         if @options[:virtwho] and @virtwho_data
           if @options[:verbose]
-            puts "VERBOSE: adding virt-who data to #{sys['name'}"
+            puts "VERBOSE: adding virt-who data to #{sys['name']}"
```
Comment (evgeni) [inline katello-attach-subscription:440]: I've split out this and the next change into https://github.com/RedHatSatellite/katello-attach-subscription/pull/42 so we can get the bugfix in quicker than the whole review :)

File: katello-attach-subscription (hunk 2 above)
Comment (mossicrue) [reply katello-attach-subscription]: doesn't become 'name="the_subname"'?
From my test it always works using that piece of code and give me problem using (name: subname) as params

Comment (evgeni) [reply katello-attach-subscription]: So this is what I get with current master:

```
irb(main):002:0> sub_name = 'something special'
=> "something special"
irb(main):003:0> puts KatelloAttachSubscription::Utils.search_args(name: sub_name)
name="something special"
=> nil
irb(main):004:0> search_params = 'name="'+sub_name+'"'
=> "name=\"something special\""
irb(main):005:0> puts KatelloAttachSubscription::Utils.search_args(search_params)
"name="something special""
=> nil
```

Note the doubled quoting in the last line.

File: katello-attach-subscription [hunk 5]
```diff
@@ -907,13 +896,37 @@ def vdcupdate()
       else
         virtualhostregex = nil
       end
+      # test all the facts
+      if @options[:verbose]
+        puts "VERBOSE: Start testing facts value for #{sys['name']}"
+      end
+      fact_test_passed = KatelloAttachSubscription::HostMatcher.match_host(sys, sub)
+      if not fact_test_passed
```
Comment (evgeni) [inline katello-attach-subscription]: `if not` can be written as `unless`

File: katello-attach-subscription [hunk 6]
```diff
@@ -852,42 +856,27 @@ def vdcupdate()
...
-          puts "FATAL ERROR: --check-density option enabled but no 'cluster' regex found in the yaml for this sub entry"
-          p sub
-          exit 5
-        end
+        # retrieve cluster name of the hypervisor
         if sys.has_key?('facts') and sys['facts'].has_key?('hypervisor::cluster')
-          this_system_cluster = sys['facts']['hypervisor::cluster']
-        else
-          this_system_cluster = 'nil'
+         this_system_cluster = sys['facts']['hypervisor::cluster']
```
Comment (evgeni) [inline katello-attach-subscription]: you lost one indentation space here

File: katello-attach-subscription [hunk 7]
```diff
@@ -852,42 +856,27 @@ def vdcupdate()
...
-          if @options[:verbose]
-            puts " VERBOSE: System cluster #{this_system_cluster} match sub cluster regex #{sub['cluster']}"
-          end
-          skip_sub = false
-        else
-          if @options[:verbose]
-            puts " VERBOSE: Skipping as cluster #{this_system_cluster} doesn't match sub cluster type #{sub['cluster']}"
-          end
+        if this_system_cluster == 'nil'
+          sys.merge!({"facts"=>{'hypervisor::cluster'=>this_system_cluster}})
```
Comment (evgeni) [inline katello-attach-subscription]: this will overwrite all other facts known about the system (as it overwrites the facts entry).

Comment (evgeni) [reply katello-attach-subscription]:
```suggestion
          sys['facts'] ||= {}
          sys['facts']['hypervisor::cluster'] = this_system_cluster
```

File: katello-attach-subscription [hunk 8]
```diff
@@ -1132,9 +1148,9 @@ def vdcupdate()
               if sub['available'] >= subscriptions_needed
                 if not @options[:noop]
                   # currently seems not working as a bug, uncomment change the code lines for a temporary workaround
-                  # @api.resource(:host_subscriptions).call(:remove_subscriptions, {:host_id => system['id'], :subscriptions => [{:id => sub['id']}]})
-                  # @api.resource(:host_subscriptions).call(:add_subscriptions, {:host_id => system['id'], :subscriptions => [{:id => sub['id'], :quantity => total_subscriptions}]})
-                  @api.resource(:host_subscriptions).call(:add_subscriptions, {:host_id => system['id'], :subscriptions => [{:id => sub['id'], :quantity => subscriptions_needed}]})
+                  @api.resource(:host_subscriptions).call(:remove_subscriptions, {:host_id => system['id'], :subscriptions => [{:id => sub['id']}]})
```
Comment (evgeni) [inline katello-attach-subscription:1152]: was this change intended? is the bug gone?

**Comment by evgeni** [comment]
two style comments, one question, and imho two bugs as noted above. the rest of the code looks fine though, so fix these and we merge

File: katello-attach-subscription (hunk 8 above)
Comment (mossicrue) [reply katello-attach-subscription:1152]: no, bug isn't gone, so the only workaround I found is to remove and reattach subs

Comment (evgeni) [reply katello-attach-subscription:1152]: ah, I see, you enable the workaround here. I miss-read, sorry.

can you add the BZ/Redmine link to the comment please?

File: katello-attach-subscription (hunk 2 above)
Comment (evgeni) [reply katello-attach-subscription]: but maybe I know what problem you were seeing, let me test something

Comment (evgeni) [reply katello-attach-subscription]: nope, not what I thought. please show me how it failed for you :)

**Review by evgeni — APPROVED** [review]
